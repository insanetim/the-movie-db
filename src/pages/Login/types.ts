import type { IUserData } from 'src/store/session/types'

export interface LoginHook {
  handleLogIn: (userData: IUserData) => Promise<void>
  loading: boolean
}
