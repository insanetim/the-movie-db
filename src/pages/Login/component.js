import { Button, Col, Form, Input, Layout, Row, Typography } from 'antd'
import { UserOutlined, LockOutlined } from '@ant-design/icons'

import useContainer from './hook'

const Login = () => {
  const { loading, handleLogIn } = useContainer()

  return (
    <div className='center'>
      <Layout>
        <Layout.Content>
          <Row justify='center'>
            <Col>
              <Form
                onFinish={handleLogIn}
                autoComplete='off'
              >
                <Typography.Title>The Movie DB</Typography.Title>
                <Form.Item
                  name='username'
                  rules={[{ required: true, message: 'Username is required' }]}
                >
                  <Input
                    prefix={<UserOutlined style={{ color: 'rgba(0,0,0,.25)' }} />}
                    placeholder='Username'
                  />
                </Form.Item>
                <Form.Item
                  name='password'
                  rules={[
                    { required: true, message: 'Password is required' },
                    { min: 4, message: 'Password must be at least 4 characters' }
                  ]}
                >
                  <Input
                    prefix={<LockOutlined style={{ color: 'rgba(0,0,0,.25)' }} />}
                    type='password'
                    placeholder='Password'
                  />
                </Form.Item>
                <Form.Item>
                  <Button
                    loading={loading}
                    type='primary'
                    htmlType='submit'
                  >
                    Log in
                  </Button>
                </Form.Item>
              </Form>
            </Col>
          </Row>
        </Layout.Content>
      </Layout>
    </div>
  )
}

export default Login
