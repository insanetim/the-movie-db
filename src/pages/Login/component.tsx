import { LockOutlined, UserOutlined } from '@ant-design/icons'
import { Button, Col, Form, Input, Layout, Row, Typography } from 'antd'
import { Helmet } from 'react-helmet'
import metaTitle from 'src/utils/helpers/metaTitle'

import useContainer from './hook'

const Login: React.FC = () => {
  const { handleLogIn, loading } = useContainer()

  return (
    <>
      <Helmet title={metaTitle('Login')} />
      <div className='center'>
        <Layout>
          <Layout.Content>
            <div className='container'>
              <Row justify='center'>
                <Col
                  lg={8}
                  md={12}
                  span={20}
                  xl={6}
                >
                  <Form
                    autoComplete='off'
                    onFinish={handleLogIn}
                  >
                    <Typography.Title>The Movie DB</Typography.Title>
                    <Form.Item
                      name='username'
                      rules={[{ message: 'Username is required', required: true }]}
                    >
                      <Input
                        placeholder='Username'
                        prefix={<UserOutlined style={{ color: 'rgba(0,0,0,.25)' }} />}
                      />
                    </Form.Item>
                    <Form.Item
                      name='password'
                      rules={[
                        { message: 'Password is required', required: true },
                        { message: 'Password must be at least 4 characters', min: 4 }
                      ]}
                    >
                      <Input
                        placeholder='Password'
                        prefix={<LockOutlined style={{ color: 'rgba(0,0,0,.25)' }} />}
                        type='password'
                      />
                    </Form.Item>
                    <Form.Item>
                      <Button
                        htmlType='submit'
                        loading={loading}
                        type='primary'
                      >
                        Log in
                      </Button>
                    </Form.Item>
                  </Form>
                </Col>
                <Col span={24}>
                  <p className='text-center'>
                    Don&apos;t have The Movie Database account?{' '}
                    <Typography.Link
                      href='https://www.themoviedb.org/signup'
                      rel='noreferrer'
                      target='_blank'
                    >
                      Sign up.
                    </Typography.Link>
                  </p>
                </Col>
              </Row>
            </div>
          </Layout.Content>
        </Layout>
      </div>
    </>
  )
}

export default Login
