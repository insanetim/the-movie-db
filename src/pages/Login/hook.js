import { useDispatch, useSelector } from 'react-redux'
import { useLocation, useNavigate } from 'react-router-dom'

import { logIn } from 'src/state/session/actions'
import { loadingSelector } from 'src/state/session/selectors'

const useContainer = () => {
  const dispatch = useDispatch()
  const loading = useSelector(loadingSelector)
  const location = useLocation()
  const navigate = useNavigate()

  const handleLogIn = userData => {
    const from = location.state?.from?.pathname || '/'
    const callback = () => navigate(from, { replace: true })
    dispatch(logIn({ userData, callback }))

    return callback
  }

  return { loading, handleLogIn }
}

export default useContainer
