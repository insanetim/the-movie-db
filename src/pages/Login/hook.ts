import type { IUserData } from 'src/store/session/types'

import { unwrapResult } from '@reduxjs/toolkit'
import { useLocation, useNavigate } from 'react-router-dom'
import { useAppDispatch, useAppSelector } from 'src/hooks/useRedux'
import { logIn } from 'src/store/session/actions'
import { loadingSelector } from 'src/store/session/selectors'

import type { LoginHook } from './types'

const useContainer = (): LoginHook => {
  const dispatch = useAppDispatch()
  const loading = useAppSelector(loadingSelector)
  const location = useLocation()
  const navigate = useNavigate()

  const handleLogIn = async (userData: IUserData) => {
    const to = location.state?.from?.pathname || '/'
    const promise = await dispatch(logIn(userData))
    const sessionId = unwrapResult(promise)
    if (sessionId) {
      navigate(to, { replace: true })
    }
  }

  return { handleLogIn, loading }
}

export default useContainer
