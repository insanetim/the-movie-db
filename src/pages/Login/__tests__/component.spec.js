import { render } from '@testing-library/react'

import Wrapper from 'src/__mocks__/wrapperMock'
import Login from '../component'

const mockedHookData = {
  loading: false,
  handleLogIn: jest.fn()
}
jest.mock('../hook', () => jest.fn(() => mockedHookData))

describe('Login component', () => {
  afterAll(() => {
    jest.unmock('../hook')
  })

  it('matches snapshot', () => {
    const { asFragment } = render(<Login />, { wrapper: Wrapper })

    expect(asFragment()).toMatchSnapshot()
  })

  it('matches snapshot with loading', () => {
    mockedHookData.loading = true
    const { asFragment } = render(<Login />, { wrapper: Wrapper })

    expect(asFragment()).toMatchSnapshot()
  })
})
