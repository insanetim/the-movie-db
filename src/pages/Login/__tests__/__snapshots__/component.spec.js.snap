// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="center"
>
  <Layout>
    <Content>
      <Row
        justify="center"
      >
        <Col>
          <ForwardRef(InternalForm)
            autoComplete="off"
            onFinish={[MockFunction]}
          >
            <ForwardRef>
              The Movie DB
            </ForwardRef>
            <InternalFormItem
              name="username"
              rules={
                [
                  {
                    "message": "Username is required",
                    "required": true,
                  },
                ]
              }
            >
              <Input
                placeholder="Username"
                prefix={
                  <ForwardRef(UserOutlined)
                    style={
                      {
                        "color": "rgba(0,0,0,.25)",
                      }
                    }
                  />
                }
              />
            </InternalFormItem>
            <InternalFormItem
              name="password"
              rules={
                [
                  {
                    "message": "Password is required",
                    "required": true,
                  },
                  {
                    "message": "Password must be at least 4 characters",
                    "min": 4,
                  },
                ]
              }
            >
              <Input
                placeholder="Password"
                prefix={
                  <ForwardRef(LockOutlined)
                    style={
                      {
                        "color": "rgba(0,0,0,.25)",
                      }
                    }
                  />
                }
                type="password"
              />
            </InternalFormItem>
            <InternalFormItem>
              <Button
                htmlType="submit"
                loading={false}
                type="primary"
              >
                Log in
              </Button>
            </InternalFormItem>
          </ForwardRef(InternalForm)>
        </Col>
      </Row>
    </Content>
  </Layout>
</div>
`;

exports[`matches snapshot with loading 1`] = `
<div
  className="center"
>
  <Layout>
    <Content>
      <Row
        justify="center"
      >
        <Col>
          <ForwardRef(InternalForm)
            autoComplete="off"
            onFinish={[MockFunction]}
          >
            <ForwardRef>
              The Movie DB
            </ForwardRef>
            <InternalFormItem
              name="username"
              rules={
                [
                  {
                    "message": "Username is required",
                    "required": true,
                  },
                ]
              }
            >
              <Input
                placeholder="Username"
                prefix={
                  <ForwardRef(UserOutlined)
                    style={
                      {
                        "color": "rgba(0,0,0,.25)",
                      }
                    }
                  />
                }
              />
            </InternalFormItem>
            <InternalFormItem
              name="password"
              rules={
                [
                  {
                    "message": "Password is required",
                    "required": true,
                  },
                  {
                    "message": "Password must be at least 4 characters",
                    "min": 4,
                  },
                ]
              }
            >
              <Input
                placeholder="Password"
                prefix={
                  <ForwardRef(LockOutlined)
                    style={
                      {
                        "color": "rgba(0,0,0,.25)",
                      }
                    }
                  />
                }
                type="password"
              />
            </InternalFormItem>
            <InternalFormItem>
              <Button
                htmlType="submit"
                loading={true}
                type="primary"
              >
                Log in
              </Button>
            </InternalFormItem>
          </ForwardRef(InternalForm)>
        </Col>
      </Row>
    </Content>
  </Layout>
</div>
`;
